scripts/evaluate_chunk_size.py

```diff
@@ -0,0 +1,148 @@
+import os
+import uuid
+import time
+
+import chromadb
+from openai import OpenAI
+
+
+EMBEDDING_COLLECTION_NAME = "learn"
+chunk_sizes = ["paragraph", "half_a_paragraph", "quarter_a_paragraph", "sentence"]
+
+chromadb_client = chromadb.HttpClient(
+    host=os.getenv("CHROMADB_HOST", "chromadb"),
+    port=8000,
+    ssl=False,
+    headers=None,
+    settings=chromadb.config.Settings(),
+    tenant=chromadb.config.DEFAULT_TENANT,
+    database=chromadb.config.DEFAULT_DATABASE,
+)
+
+openai_client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
+chromadb_client = chromadb.HttpClient(
+    host=os.getenv("CHROMADB_HOST", "chromadb"),
+    port=8000,
+    ssl=False,
+    headers=None,
+    settings=chromadb.config.Settings(),
+    tenant=chromadb.config.DEFAULT_TENANT,
+    database=chromadb.config.DEFAULT_DATABASE,
+)
+
+# Questions to ask about the text
+f_query = open("../data/RAG_chunk_evaluation_questions.txt", "r")
+my_queries = f_query.read().split("\n")
+f_query.close()
+
+unique_file_id = str(int(time.time()))
+
+for chunk_size_desc in chunk_sizes:
+
+    print(chunk_size_desc)
+
+    # a file to store responses
+    save_path = "evaluate_chunk_size_results"
+    if not os.path.exists(save_path):
+        os.makedirs(save_path)
+    f_response = open(
+        save_path + "/response_" + unique_file_id + "_" + chunk_size_desc + ".txt", "w"
+    )
+    f_response.writelines([chunk_size_desc, "\n"])
+
+    # read and chunk the text
+    f = open("../data/chapterXX.txt", "r")
+    chunk_sizes = ["paragraph", "half_a_paragraph", "quarter_a_paragraph", "sentence"]
+    if chunk_size_desc == "paragraph":
+        my_documents = f.read().split("\n")
+        my_ids = [f"{uuid.uuid4().hex}" for _ in range(len(my_documents))]
+    elif chunk_size_desc == "half_a_paragraph":
+        my_documents_tmp = f.read().split("\n")
+        my_documents = []
+        for i in range(
+            len(my_documents_tmp)
+        ):  # Computationally heavy. Let's look for a python library.
```

Comment:
having this comment here is misleading because it indicates the for-loop itself is a heavy operation though it is not (len(list) only takes O(1) in Python because Python list holds indexes).
What exactly is a computationally heavy statement?

Reply:
using `enumerate` might make this logic simpler

Reply:
I see...

Let's say `m` is the number of lines (documents) and `n` is the length of a line
Because creating a new string with slices is basically copying a string to a new address, its time complexity is O(n). No matter how many times we do (2 or 4), O((2 or 4) * n) = O(n). Therefore, the total time complexity of this block is `O(mn)`. Because `m` and `n` here are relatively small numbers, I don't think this code block is a heavy operation. It should finish in a short time.

and my gut feeling is that `O(mn)` is the optimal (with simple implementations). We need to iterate through all the lines anyway `O(m)`, and halving (actually creating in Python) a string of length = `n` costs `O(n)`.

Reply:
Ah I see, it is good already then😌👍
Thank you for explaining! I will erase the comment line